### PR TITLE
[CENPLAT-23916] Add Python bindings using pyo3

### PR DESCRIPTION
**Pull request by jsutton24**

### CODE REVIEW

File: omf-python/docs/source/conf.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Configuration file for the Sphinx documentation builder.
+#
+# For the full list of built-in configuration values, see the documentation:
+# https://www.sphinx-doc.org/en/master/usage/configuration.html
+
+import omf_python
+
+project = 'OMF Rust Python API'
+copyright = '2024, Seequent'
+author = 'Catalyst IT'
```
Comment (jsutton24) [inline omf-python/docs/source/conf.py]: Note that for now we’ve attributed the copyright to Seequent and the authorship to Catalyst, but we can change this if needed.

File: omf-python/src/file/reader.rs [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+/// - Optional: adjust the limits with `reader.set_limits(...)`.
+/// - Read the project from the file with `reader.project()`.
+/// - Iterate through the project's contents to find the elements and attributes you want to load.
+/// - For each of those items load the array or image data.
+///
+/// **Warning:**
+///     When loading arrays and images from OMF files, beware of "zip bombs"
+///     where data is maliciously crafted to expand to an excessive size when decompressed,
+///     leading to a potential denial of service attack.
+///     Use the limits provided check arrays sizes before allocating memory.
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/file/reader.rs]: Might be an error in where this was copied from.
```suggestion
///     Use the limits provided and check arrays sizes before allocating memory.
```

File: omf-python/src/file/reader.rs [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    #[pyo3(get, set)]
+    pub json_bytes: Option<u64>,
+    #[pyo3(get, set)]
+    pub image_bytes: Option<u64>,
+    #[pyo3(get, set)]
+    pub image_dim: Option<u32>,
+    #[pyo3(get, set)]
+    pub validation: Option<u32>,
+}
+
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/file/reader.rs:38]: It can be tidier to provide `From` implementations for the conversions back and forth from the wrapped types. This code is untested:
```suggestion
impl From<Limits> for PyLimits {
    fn from(limits: Limits) -> Self {
        Self {
            json_bytes: limits.json_bytes,
            image_bytes: limits.image_bytes,
            image_dim: limits.image_dim,
            validation: limits.validation,
        }
    }
}

impl From<PyLimits> for Limits {
    fn from(py_limits: PyLimits) -> Self {
        Self {
            json_bytes: py_limits.json_bytes,
            image_bytes: py_limits.image_bytes,
            image_dim: py_limits.image_dim,
            validation: py_limits.validation,
        }
    }
}
```
Then anywhere that needs to convert between the two can just use `limits.into()`.

File: omf-python/src/file/reader.rs [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+    fn project(&self) -> PyResult<(PyProject, Vec<PyProblem>)> {
+        let (project, problems) = self.0.project().map_err(OmfException::py_err)?;
+
+        let problems_array: Vec<PyProblem> =
+            problems.iter().map(|e| PyProblem(e.clone())).collect();
+        Ok((PyProject(project), problems_array))
+    }
+
+    /// Read a Scalar array.
+    pub fn array_scalars(&self, array: &PyScalarArray) -> PyResult<Vec<f64>> {
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/file/reader.rs]: I believe the automatic conversion from `pyo3` will convert a `Vec` into a Python `list`. This will be too slow for real data, as each list item will need to be a new Python object, allocating tens of millions or more objects for a large array.

The conversion will need to use the [`numpy`](https://crates.io/crates/numpy) crate to instead return NumPy arrays to Python. Using the [`ndarray`](https://crates.io/crates/ndarray/) crate in this wrapper code might make it easier. Ideally this would preserve `f32` vs `f64` choice of the data in the OMF file, reducing memory usage and preserving the precision of the file so callers can make informed choices around it.

The complicated part will be attribute array types with nulls. While 1D arrays cold be returned as Numpy masked arrays, 2D arrays like vectors and colours won't work like that, because the mask is on each element rather than each row. The best option is probably to return an `(array, mask)` tuple for all nullable types rather than using masked arrays directly.

File: omf-python/src/geometry.rs [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+#[gen_stub_pyclass]
+#[pyclass(name = "PointSet")]
+/// Point set geometry.
+pub struct PyPointSet(pub PointSet);
+
+#[gen_stub_pymethods]
+#[pymethods]
+impl PyPointSet {
+    #[getter]
+    fn origin(&self) -> [f64; 3] {
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/geometry.rs]: It may be better to return NumPy arrays for these 3D vectors as well, for ease of use inside Python code.

As-is this will return a Python list, not even a tuple, which may be a bit confusing. A smaller change would be to return `(f64, f64, f64)` so the Python object will be a tuple.

File: omf-python/src/colormap.rs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        match self.0 {
+            NumberColormap::Continuous { range, .. } => match range {
+                NumberRange::Float { min, max, .. } => Ok((min, max).into_py(py)),
+                NumberRange::Integer { min, max, .. } => Ok((min, max).into_py(py)),
+                NumberRange::Date { min, max, .. } => Ok((min, max).into_py(py)),
+                NumberRange::DateTime { min, max, .. } => Ok((min, max).into_py(py)),
+            },
+            _ => Err(PyValueError::new_err(
+                "NumberColormap variant is not supported",
+            )),
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/colormap.rs]: Other similar code is just panicking in this case, which I think is better, as it indicates that reaching here requires a programming error in these bindings. Remaining consistent is better too.
```suggestion
            _ => unreachable!(),
```

File: docs/start.md [hunk 7]
```diff
@@ -36,8 +36,78 @@ See the [C API documentation](c/index.md).
 
 ## Python API
 
-Not yet written.
-
+This is a work in progress. To contribute to building Python bindings, in the project root:
```
Comment (Tim-Evans-Seequent) [inline docs/start.md]: Can you add a list here of what isn't wrapped yet?

File: omf-python/src/element.rs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        match &self.0.geometry {
+            Geometry::PointSet(point_set) => Ok(PyPointSet(point_set.clone()).into_py(py)),
+            Geometry::LineSet(line_set) => Ok(PyLineSet(line_set.clone()).into_py(py)),
+            Geometry::Surface(surface) => Ok(PySurface(surface.clone()).into_py(py)),
+            Geometry::GridSurface(grid_surface) => {
+                Ok(PyGridSurface(grid_surface.clone()).into_py(py))
+            }
+            _ => Err(OmfNotSupportedException::new_err(
+                "Geometry type not supported",
+            )),
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/element.rs]: It would be nice to include the name of the unsupported geometry, or just it's Debug string:
```suggestion
            unsupported => Err(OmfNotSupportedException::new_err(
                format!("Geometry type not supported for {unsupported:?}"),
            )),
```

File: omf-python/src/grid.rs [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+pub struct PyGrid2Regular(Grid2);
+
+impl TryFrom<Grid2> for PyGrid2Regular {
+    type Error = ();
+
+    fn try_from(value: Grid2) -> Result<Self, Self::Error> {
+        match value {
+            // A Regular grid can be converted to PyGrid2Regular.
+            Grid2::Regular { .. } => Ok(Self(value)),
+            _ => Err(()),
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/grid.rs]: As in other places, it's probably fine to use `From` with an `unreachable!` panic, if there is other code making this choice.

File: docs/start.md (hunk 7 above)
Comment (andrew-catalyst) [inline docs/start.md]: Can we move our changes here somewhere else (like a python-bindings.md, or as the README.md in the omf-python folder) since this is developer documentation rather than for API users.

Maybe put in a place holder link to the python docs, though it wont work currently.

File: omf-python/src/attribute.rs [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+    /// Array with `Index` type storing the category indices.
+    ///
+    /// Values are indices into the `names` array, `colors` array, and any sub-attributes,
+    /// and must be within range for them.
+    fn values(&self) -> PyResult<PyIndexArray> {
+        match &self.0 {
+            AttributeData::Category { values, .. } => Ok(PyIndexArray(values.clone())),
+            _ => Err(OmfNotSupportedException::new_err(
+                "AttributeData variant is not supported",
+            )),
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/attribute.rs]: Again, I think it's tidier for this to panic.

Also, this pattern is repeated a lot here. Using a macro would make it more readable. This function could be reduced to:
```rust
fn values(&self) -> PyIndexArray {
    attribute_data_field!(self, Category::values).clone()
}
```
With a macro like this (untested):
```rust
macro_rules! attribute_data_field {
    ($self:ident, $variant:ident :: $name:ident) => {
        match &$self.0 {
            AttributeData::$variant { $field, .. } => &$field,
            _ => unreachable!(),
        }
    };
}
```

File: omf-python/src/geometry.rs [hunk 11]
```diff
@@ -59,16 +59,23 @@ pub struct PySurface(pub Surface);
 #[pymethods]
 impl PySurface {
     #[getter]
+    /// Origin of the surface relative to the project coordinate reference system.
```
Comment (andrew-catalyst) [inline omf-python/src/geometry.rs:63]: Just thought I should mention somewhere, a caveat of the current way the python docstrings ares generated is that `pyo-stub-gen` doesn't translate getters to anything at the moment. I think because as it turns them into variables in the generated `omf_python.pyi` and in python variables don't have docstrings (at least, I think I tried adding them manually and didn't work).

If they changed it to instead generate python `@property` decorated methods I think they would have a place to go.

We could maintain `omf_python.pyi` manually but then its more likely to get out of sync with the code easily.

**Review by Tim-Evans-Seequent — CHANGES_REQUESTED** [review]

File: omf-python/src/geometry.rs [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+                .into_py(py),
+            omf::Grid2::Tensor { .. } => PyGrid2Tensor::try_from(self.0.grid.clone())
+                .expect("conversion from Tensor should succeed")
+                .into_py(py),
+        }
+    }
+
+    /// Array with `Scalar` type storing the offset of each grid vertex from the place.
+    /// Heights may be positive or negative. Will be absent from flat 2D grids.
+    fn heights(&self) -> Option<PyScalarArray> {
```
Comment (andrew-catalyst) [inline omf-python/src/geometry.rs:111]: To be consistent with the other getter methods that return an `Array` type this should probably be a getter also. I misunderstood what it was doing when I first reviewed it.

File: docs/start.md (hunk 7 above)
Comment (jsutton24) [reply docs/start.md]: I’ve moved this text into `omf-python/README.md`.

File: omf-python/src/file/reader.rs (hunk 4 above)
Comment (jsutton24) [reply omf-python/src/file/reader.rs]: We’re now returning numpy arrays for the methods with more conventional return types. I didn’t notice much difference in execution time for a file with an artificially large number of vertices, but I observed a sixfold reduction in the size of the vertex array in memory.

For return types without an analogue in numpy, such as `Vec<Option<u32>>`, we’ll probably have to use the the impromptu masking approach. (The numpy crate doesn’t support masked arrays yet.)

Comment (Tim-Evans-Seequent) [reply omf-python/src/file/reader.rs]: What counts as artificially large in this case? At the top end, a block model with 100 million sub-blocks and 20+ attributes wouldn't be unheard-of.

Comment (jsutton24) [reply omf-python/src/file/reader.rs]: That’s true. In this case “an artificially large number” was one million.

Comment (Tim-Evans-Seequent) [reply omf-python/src/file/reader.rs]: Yeah, that's not the point where you'd see the difference I think. One million to me would just be a fairly normal mesh.

**Comment by jsutton24** [comment]
I’ve rewritten a number of the `Reader` methods to return numpy masked arrays, as that turned out to be fairly straightforward, and presumably more performant than Python lists and easier to use for consumers of the Python API. @Tim-Evans-Seequent is that a reasonable approach, or would you rather we returned Python lists of `(array, mask)` tuples?

**Comment by jsutton24** [comment]
> I’ve rewritten a number of the `Reader` methods to return numpy masked arrays, as that turned out to be fairly straightforward, and presumably more performant than Python lists and easier to use for consumers of the Python API. @Tim-Evans-Seequent is that a reasonable approach, or would you rather we returned Python lists of `(array, mask)` tuples?

Any opinion on this?

File: omf-python/src/file/reader.rs [hunk 13]
```diff
@@ -132,84 +166,235 @@ impl PyReader {
...
+
+                let shape = (vec.len(), 3);
+
+                // TODO: remove unwrap
+
+                Ok(PyArray::from_owned_array_bound(
+                    py,
+                    Array::from_shape_vec(shape, vec.into_flattened()).unwrap(),
+                )
+                .into_py(py))
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/file/reader.rs]: Don't repeat yourself. Module level-functions for converting flat, wide (like this one), nullable-flat, and nullable-wide arrays should tidy this up.

File: omf-python/src/file/reader.rs [hunk 14]
```diff
@@ -132,84 +166,235 @@ impl PyReader {
...
+        );
+
+        let shape = (mask.len(), 4);
+        // TODO: remove unwrap
+        let mask = PyArray::from_owned_array_bound(
+            py,
+            Array::from_shape_vec(shape, mask.into_flattened()).unwrap(),
+        );
+
+        Ok(masked_array(py, colors, mask)?.downcast_into()?)
```
Comment (Tim-Evans-Seequent) [inline omf-python/src/file/reader.rs]: I suspect that a masked array here won't be all that useful to the caller. NumPy masked arrays mask each element, but for this and similar arrays from OMF it's each row that can be null or not. In my experience doing any processing with a masked array like this is a pain, as operations on rows didn't skip null rows like I expected them to.

This will be of a problem for writing too, where the Python bindings would need to check that masked arrays are following the right pattern.

Returning and accepting an `(array, mask)` tuple might be more useful. For colors the array would have shape (N, 4) and the mask just (N,). The caller could still create a masked array if they want one, or compress the colors with `array.compress(mask, axis=0)` which is a pain with a 2D masked array.

It's been a while since I looked at this though, and I can't remember all the details. Maybe try out a few operations like mapping RGB to sRGB or HSV and see if it works well. This only affects colors and vectors in OMF, but I would prefer consistency across all nullable array types.

**Review by Tim-Evans-Seequent — CHANGES_REQUESTED** [review]

**Comment by Tim-Evans-Seequent** [comment]
> > I’ve rewritten a number of the `Reader` methods to return numpy masked arrays, as that turned out to be fairly straightforward, and presumably more performant than Python lists and easier to use for consumers of the Python API. @Tim-Evans-Seequent is that a reasonable approach, or would you rather we returned Python lists of `(array, mask)` tuples?
> 
> Any opinion on this?

Sorry, Github wasn't showing me the new commits through the normal workflow. That looks about right, though I do prefer (array, mask) due to the 2D array types. I've struck problems with 2D masks in the past.

File: omf-python/src/file/reader.rs (hunk 13 above)
Comment (jsutton24) [reply omf-python/src/file/reader.rs]: Yep, I’m planning to move all the repeated code into methods.

File: omf-python/src/file/reader.rs (hunk 14 above)
Comment (jsutton24) [reply omf-python/src/file/reader.rs]: Okay, returning a pair of numpy arrays makes sense.

File: omf-python/src/file/reader.rs [hunk 15]
```diff
@@ -261,53 +460,85 @@ impl PyReader {
...
 
     /// Read a Boundary array.
-    pub fn array_boundaries(&self, array: &PyBoundaryArray) -> PyResult<Vec<f64>> {
-        let numbers_f64 = self
+    pub fn array_boundaries<'py>(
+        &self,
+        py: Python<'py>,
+        array: &PyBoundaryArray,
+    ) -> PyResult<Bound<'py, PyAny>> {
+        // TODO: using value() to get the boundary value is lossy.
```
Comment (jsutton24) [inline omf-python/src/file/reader.rs]: Any thoughts on how we ought to transform arrays of Boundary values to numpy arrays? The current way loses the distinction between than `Less` and `LessEqual` values, which is presumably important.

File: omf-python/tests/test_number_attribute.py [hunk 16]
```diff
@@ -53,22 +62,28 @@ def test_should_return_discrete_colormap_boundaries(self) -> None:
         )
         actual_boundaries = self.reader.array_boundaries(boundary_array)
 
-        expected_boundaries = [946688400.0, 946692000.0, 946695600.0]
-        self.assertEqual(actual_boundaries, expected_boundaries)
+        # TODO: check these values
+        expected_boundaries = numpy.array(
+            [946688400000000, 946692000000000, 946695600000000]
+        )
```
Comment (jsutton24) [inline omf-python/tests/test_number_attribute.py]: Also need to determine why the type‐aware `array_boundaries()` returns different values.

File: omf-python/src/file/reader.rs (hunk 15 above)
Comment (Tim-Evans-Seequent) [reply omf-python/src/file/reader.rs]: It should probably be returning a list of tuples or objects rather than an array. It would be surprising for there to be more than a hundred of them, so performance will be fine.

File: omf-python/tests/test_number_attribute.py (hunk 16 above)
Comment (jsutton24) [reply omf-python/tests/test_number_attribute.py]: I think it’s because these are datetimes that were being returned with microseconds precision rather than seconds precision.

File: omf-python/src/file/reader.rs (hunk 15 above)
Comment (jsutton24) [reply omf-python/src/file/reader.rs]: We’re now returning a list of `(value, BoundaryType)` tuples.

File: omf-python/src/grid.rs (hunk 9 above)
Comment (jsutton24) [reply omf-python/src/grid.rs]: I think a panicking `From` implementation would be generally unexpected. [The `From` documentation](https://doc.rust-lang.org/std/convert/trait.From.html#when-to-implement-from) also recommends against implementations that panic.

Comment (Tim-Evans-Seequent) [reply omf-python/src/grid.rs]: In a typical situation I would agree, but I feel like this isn't one. If it was, using `()` as the error type probably wouldn't be great either.

For self-contained code where the only way it can go wrong is if that code is wrong, a panic is often more clear. Here it seems like everything that calls this immediately chains to `.expect(...)` and it would be more obvious to just panic right away and avoid the extra code.

I don't feel too strongly about it in this specific case, but there was a pattern of unnecessary error handling in other places as well.

File: docs/start.md (hunk 7 above)
Comment (jsutton24) [reply docs/start.md]: I’m not sure. I think that most things should be wrapped?

**Comment by jsutton24** [comment]
Is there anything else to prevent merging this?

**Review by Tim-Evans-Seequent — APPROVED** [review]

File: omf-python/tests/test_number_attribute.py (hunk 16 above)
Comment (Tim-Evans-Seequent) [reply omf-python/tests/test_number_attribute.py]: Ideally this would return an array with dtype "datetime64[us]", but that might be a pain. It's fine as-is if that's easier.

**Review by andrew-catalyst — APPROVED** [review]

File: omf-python/tests/test_number_attribute.py (hunk 16 above)
Comment (jsutton24) [reply omf-python/tests/test_number_attribute.py]: Currently I’ve got it returning tuples with [datetime](https://docs.python.org/3/library/datetime.html) objects.
